### ISSUE TITLE
`AppDaemon.sun_down` has timezone issues

### ISSUE THREAD

**Issue by puddly** [home-assistant/appdaemon#250]
I've been having some problems with automations relying on `AppDaemon.sun_down`. After some debugging, both `self.AD.sun['next_rising']` and `self.AD.sun['next_setting']` appear to be off by a day:

    In [21]: now = datetime.now(pytz.UTC)

    In [22]: str(self.AD.sun['next_rising'] - now)
    Out[22]: '14:10:31.195186'

    In [23]: str(self.AD.sun['next_setting'] - now)
    Out[23]: '1 day, 1:23:54.195186'

Home Assistant correctly reports the next sunset to be in about three hours, yet AppDaemon is off by a day, which flips the `sun_up` and `sun_down` automations. Looking at the code, it seems like the issue stems from these two lines in `appdaemon.AppDaemon`:

    def update_sun(self):
        # now = datetime.datetime.now(self.tz)
        now = self.tz.localize(self.get_now())

The commented out line correctly reports my time, while the second one just attaches my timezone to the current UTC time, which is wrong.

Is this a configuration issue on my end?

**Comment by acockburn** [comment]
The commented out line was replaced by the current one so that time travel would work properly - I'm not sure why it's broken. What timezone are you using?

**Comment by puddly** [comment]
I'm using `America/New_York` in my Home Assistant configuration file but changing it to another timezone doesn't make a difference.

Am I misreading the intent of the code? `self.get_now()` returns a timezone-unaware `datetime` object that implicitly uses UTC and then it's just given my local timezone, which results in `now` containing the wrong time.

For calculating the time the sun will rise at a particular point, I don't think the local timezone needs to be involved anywhere beyond the `astral.Location` initializer in `AppDaemon.init_sun`. Will the following change work?

    now = pytz.utc.localize(self.get_now())

I think it'd be better to have `AppDaemon.get_now` return a timezone-aware datetime object.

**Comment by acockburn** [comment]
What API calls are you using? My code shows all things working as expected, when using the published interfaces. and if there were an issue here the scheduler would be out of whack for sunrise and sunset, and I'm not having any issues with that.

```python
  def get_sun_info(self):
    self.log("Current sun state: Up = {}, Down = {}".format(self.sun_up(), self.sun_down()))
    self.log("Next sunrise: {}".format(self.sunrise()))
    self.log("Next sunset: {}".format(self.sunset()))
```

gives:

```
2018-03-16 15:48:55.365078 INFO AppDaemon: Initializing app dep1 using class Sun from module sun
2018-03-16 15:48:55.368289 INFO dep1: Current sun state: Up = True, Down = False
2018-03-16 15:48:55.370860 INFO dep1: Next sunrise: 2018-03-17 07:02:44
2018-03-16 15:48:55.374357 INFO dep1: Next sunset: 2018-03-16 19:00:20
```

Which is correct.

What update_sun() is actually doing is coming up with timezone aware datetime objects that are in the UTC timezone.

**Comment by puddly** [comment]
I was using `self.sun_up()` and `self.sun_down()` in my automations. If I use the following code:

    from appdaemon.plugins.hass.hassapi import Hass


    class Test(Hass):
        def initialize(self):
            self.log("Current sun state: Up = {}, Down = {}".format(self.sun_up(), self.sun_down()))
            self.log("Next sunrise: {}".format(self.sunrise()))
            self.log("Next sunset: {}".format(self.sunset()))

I get the following output:

    2018-03-16 20:09:52.076777 INFO light_manager: Current sun state: Up = False, Down = True
    2018-03-16 20:09:52.079008 INFO light_manager: Next sunrise: 2018-03-17 11:44:21
    2018-03-16 20:09:52.081121 INFO light_manager: Next sunset: 2018-03-17 23:42:47

I'm using `appdaemon==3.0.0b4`.

**Comment by acockburn** [comment]
I'm in the same timezone as you and it isn't 20:09 here yet, so something is wrong with your timezone.

By default, AD will take the timezone from HASS, but you can override it in the appdaemon.yaml file - are you doing that by any chance?

**Comment by acockburn** [comment]
Those sunrise and sunset times, as well as your current timestamps, if you ran this just now all point to your timezone being set to UTC

**Comment by acockburn** [comment]
Your times are different to min but that is probably because you are close to the west of the timezone, and I am close to the East.

**Comment by puddly** [comment]
No, my `appdaemon.yaml` file is pretty simple:

```
secrets: .homeassistant/secrets.yaml

appdaemon:
    logfile: STDOUT
    errorfile: STDERR
    threads: 10

    plugins:
        hass:
            type: hass
            ha_url: http://127.0.0.1:8123
```

I set the timezone in Home Assistant as follows:

```
homeassistant:
    name: Home
    time_zone: America/New_York
```

`self.AD.tz` is `<DstTzInfo 'America/New_York' LMT-1 day, 19:04:00 STD>` for me. Home Assistant correctly displays all of the time information and runs on the same device in a separate virtual environment.

**Comment by acockburn** [comment]
Something is strange here - my  self.AD.tz is 'America/New_York` which isn't even a TzInfo object.

What python version and platform are you running on?

**Comment by puddly** [comment]
I'm running appdaemon in its own separate virtual environment on an ODROID XU4 running Arch Linux ARM:

```$ python
Python 3.6.4 (default, Jan 25 2018, 02:05:01)
[GCC 7.2.1 20180116] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>>
$ pip freeze
aiohttp==2.3.10
aiohttp-jinja2==0.15.0
appdaemon==3.0.0b4
astral==1.6
async==0.6.2
async-timeout==2.0.0
attrs==17.4.0
bcrypt==3.1.4
certifi==2018.1.18
cffi==1.11.5
chardet==3.0.4
configparser==3.5.0
daemonize==2.4.7
decorator==4.2.1
feedparser==5.2.1
idna==2.6
idna-ssl==1.0.0
ipython==6.2.1
ipython-genutils==0.2.0
iso8601==0.1.12
jedi==0.11.1
Jinja2==2.10
MarkupSafe==1.0
multidict==4.1.0
parso==0.1.1
pexpect==4.4.0
pickleshare==0.7.4
prompt-toolkit==1.0.15
ptyprocess==0.5.2
pycparser==2.18
Pygments==2.2.0
pytz==2018.3
PyYAML==3.12
requests==2.18.4
simplegeneric==0.8.1
six==1.11.0
sseclient==0.0.18
traitlets==4.3.2
urllib3==1.22
voluptuous==0.11.1
wcwidth==0.1.7
websocket-client==0.47.0
yarl==1.1.0
```

**Comment by acockburn** [comment]
I'm running pytz v2017.2 - maybe you could try downgrading to that?

**Comment by puddly** [comment]
Downgrading to `pytz==2017.2` only changes the way the timezone object is displayed:

```
2018-03-16 20:52:54.682989 INFO light_manager: Current sun state: Up = False, Down = True
2018-03-16 20:52:54.685273 INFO light_manager: Next sunrise: 2018-03-17 11:44:21
2018-03-16 20:52:54.687360 INFO light_manager: Next sunset: 2018-03-17 23:42:47
2018-03-16 20:52:54.689445 INFO light_manager: Timezone is America/New_York
```

**Comment by acockburn** [comment]
I'm not sure what is going on here - I have a similar setup to you in terms of timezones and config and it's all working fine for me, has been since I first wrote AD. I tried your proposed changed, and that works fine for me too, does it work for you? I'm loathe to change it though until I understand what is causing this.

**Comment by puddly** [comment]
I think the root issue is `datetime.now()` using the system timezone (which for me is UTC) and the code is assuming that it will be the same as the timezone pulled from HASS (which for me is America/New_York). Using `datetime.utcnow()` might be a better solution, since it doesn't depend on the system timezone.

**Comment by acockburn** [comment]
OK, so I set my machine to UTC, and it looks like HASS gets confused as well:

![image](https://user-images.githubusercontent.com/9850788/37608183-4577aa90-2b70-11e8-970c-e057f369a50d.png)

Is this not what you are seeing?

If you would create a PR to convert to timezone aware objects I would definitely look at including it as I see the problem. My only worry would be  backward compatibility for apps - any apps using self.get_date() for instance might break. Do you have any thoughts on that?

**Comment by chkorn** [comment]
I can reproduce this issue. Details that might be useful: 

* My Timezone: Europe/Berlin (set in HASS)
* System (Docker-Host) is UTC 
* AD uses HASS TZ information
* Currently it's 21:25 (ish). 
* Sun is already down since 20:16

# Sun-Information from HASS 
![image](https://user-images.githubusercontent.com/280817/39148565-ea3f7692-473c-11e8-8e44-039468a3260a.png)

# Sun-Information from AD 
![image](https://user-images.githubusercontent.com/280817/39148468-a947f88a-473c-11e8-96cc-a4bedf1e7627.png)

--

I would also suggest converting all datetime objects into tz-aware objects to address this issue as @puddly suggested.

**Comment by puddly** [comment]
@chkorn If you're interested in helping finish up the changes, I've converted the entire codebase (you can see the [WIP here](https://github.com/puddly/appdaemon/tree/datetime-fixes)) but there are some major breaking bugs for now that I haven't had time to work through.

**Comment by chkorn** [comment]
I'll look into it! Since I just started using AD I don't have a very large set of apps where I can stumble across bugs. Maybe I'll create a few tests first.

**Comment by simonvanderveldt** [comment]
@puddly Have you been able to make any progress on this? I'm running into odd sunrise behavior as well, next sunrise is set before the current time, which means `sun_down` behavior is broken.

```
2018-10-17 06:59:31.890317 INFO motion_to_lights: time now 2018-10-17 06:59:31.890214
2018-10-17 06:59:31.890449 INFO motion_to_lights: sunrise 2018-10-17 06:09:34
2018-10-17 06:59:31.890565 INFO motion_to_lights: sunset 2018-10-17 16:44:17
```

Values in HA seem to be correct:
```
next_rising: 2018-10-18T06:11:20+00:00
next_setting: 2018-10-17T16:44:17+00:00
```

**Comment by simonvanderveldt** [comment]
@puddly Thanks for the answer!

> but it might be easier to just override the `sun_{up,down}` functions to properly handle timezones in your `Hass` subclass.

Yeah, that's what I'm doing now, quiet an easy fix this way :)

<offtopic> Out of interest, what replacement for Home Assistant + AppDaemon have you switched to?

**Comment by acockburn** [comment]
@simonvanderveldt `Yeah, that's what I'm doing now, quiet an easy fix this way :)`

Can you share what you are doing to help me get a handle on this and potentially fix it?

**Comment by simonvanderveldt** [comment]
@acockburn Sorry, I've removed any dependency on time (and by extension the sun's state) so don't have thate code anymore.
I believe it was effectively just getting the current time and comparing that vs the sun times from HA.

**Comment by acockburn** [comment]
The good news is that I’ve spent the last 2 days rewriting the scheduler to use UTC tz date aware date time objects. Still testing but I think this will resolve this issue.

**Comment by acockburn** [comment]
OK, fix is looking good to me and available in the dev branch. It will be released in 4.0. Closing this, but happy to repoen if further issues are found in this area.
